CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/utils/MediaStoreUtils.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.example.cameraxbasic.utils
+
+import android.content.ContentUris
+import android.content.Context
+import android.database.Cursor
+import android.net.Uri
+import android.os.Build
+import android.provider.MediaStore
+import androidx.core.net.toUri
+import java.io.File
+
+
+/**
+ * A utility class for accessing this app's photo storage.
+ *
+ * Since this app doesn't request any external storage permissions, it will only be able to access
+ * photos taken with this app. If the app is uninstalled, the photos taken with this app will stay
+ * on the device, but reinstalling the app will not give it access to photos taken with the app's
+ * previous instance. You can request further permissions to change this app's access. See this
+ * guide for more: https://developer.android.com/training/data-storage.
+ */
+class MediaStoreUtils(private val context: Context) {
+
+    // Suppress DATA index deprecation warning since we need the file location for the Glide library
+    @Suppress("DEPRECATION")
+    private val imageDataColumnIndex = MediaStore.Images.Media.DATA
+    private val imageIdColumnIndex = MediaStore.Images.Media._ID
+    val mediaStoreCollection: Uri? = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL)
+    } else {
+        context.getExternalFilesDir(null)?.toUri()
+    }
+
+    private fun getMediaStoreImageCursor(mediaStoreCollection: Uri): Cursor? {
+        val projection = arrayOf(imageDataColumnIndex, imageIdColumnIndex)
+        val sortOrder = "DATE_ADDED DESC"
+        return context.contentResolver.query(
+            mediaStoreCollection,  projection, null, null, sortOrder)
+    }
+
+    fun getLatestImageFilename(): String? {
+        if (mediaStoreCollection == null) return null
+
+        val cursor = getMediaStoreImageCursor(mediaStoreCollection)
+        if (cursor?.moveToFirst() != true) return null
+
+        val filename = cursor.getString(cursor.getColumnIndexOrThrow(imageDataColumnIndex))
```

Comment:
If the operation throws an exception then this means the cursor wouldn't be closed I think
Maybe it's safer to wrap in a try / finally block?

Reply:
Thanks for catching! I did that, and then Android Studio suggested `use`, which looks even cleaner.